### PR TITLE
editor: toggle style refinements

### PR DESCRIPTION
**Pull request by tvanderstad**
- fixes https://github.com/lockbook/lockbook/issues/2079
- fixes an issue where toggling style with empty selection in a nonempty document would do nothing
- fixes an issue where toggling the current style at the end of the styled region would insert characters instead of advancing to after the region
- fixes anything else using range binary search (implemented since last release)
- improves toggle style when selection bounds are in syntax sequences

### CODE REVIEW

File: libs/editor/egui_editor/src/bounds.rs [hunk 1]
```diff
@@ -571,36 +571,38 @@ impl<Range: RangeExt<DocCharOffset>> RangesExt for Vec<Range> {
     ) -> (usize, usize) {
         match self.binary_search_by(|range| {
             if offset < range.start() {
-                Ordering::Less
+                Ordering::Greater
```
Comment (tvanderstad) [inline libs/editor/egui_editor/src/bounds.rs:574]: fix: these were all backwards

File: libs/editor/egui_editor/src/bounds.rs [hunk 2]
```diff
@@ -571,36 +571,38 @@ impl<Range: RangeExt<DocCharOffset>> RangesExt for Vec<Range> {
...
             } else if offset == range.start() {
-                if start_inclusive {
+                if offset == range.end() && !end_inclusive {
```
Comment (tvanderstad) [inline libs/editor/egui_editor/src/bounds.rs:576]: fix: handling of empty ranges

File: libs/editor/egui_editor/src/bounds.rs [hunk 3]
```diff
@@ -571,36 +571,38 @@ impl<Range: RangeExt<DocCharOffset>> RangesExt for Vec<Range> {
...
-                Ordering::Greater
+                Ordering::Less
             } else {
                 unreachable!()
             }
         }) {
             Ok(idx) => {
                 let mut start = idx;
-                while start > 0 && self[idx - 1].contains(offset, start_inclusive, end_inclusive) {
+                while start > 0 && self[start - 1].contains(offset, start_inclusive, end_inclusive)
```
Comment (tvanderstad) [inline libs/editor/egui_editor/src/bounds.rs:599]: fix: using wrong variable

File: libs/editor/egui_editor/src/bounds.rs [hunk 4]
```diff
@@ -571,36 +571,38 @@ impl<Range: RangeExt<DocCharOffset>> RangesExt for Vec<Range> {
...
+                {
                     start -= 1;
                 }
 
                 let mut end = idx;
-                while end < self.len() - 1
-                    && self[end + 1].contains(offset, start_inclusive, end_inclusive)
+                while end < self.len() && self[end].contains(offset, start_inclusive, end_inclusive)
```
Comment (tvanderstad) [inline libs/editor/egui_editor/src/bounds.rs:605]: fix: off by one error

File: libs/editor/egui_editor/src/input/mutation.rs [hunk 5]
```diff
@@ -512,29 +512,25 @@ fn apply_style(
 
     // modify head/tail for nodes containing cursor start and cursor end
     let mut last_start_ancestor: Option<usize> = None;
-    if start_range.range_type == AstTextRangeType::Text {
```
Comment (tvanderstad) [inline libs/editor/egui_editor/src/input/mutation.rs:515]: fix: better toggle style handling when cursor bounds are in syntax sequences

File: libs/editor/egui_editor/src/input/mutation.rs [hunk 6]
```diff
@@ -617,6 +627,36 @@ fn detail_ast_node(node_idx: usize, ast: &Ast, mutation: &mut Vec<SubMutation>)
     mutation.push(SubMutation::Insert { text: "".to_string(), advance_cursor: false });
 }
 
+fn adjust_for_whitespace(
```
Comment (tvanderstad) [inline libs/editor/egui_editor/src/input/mutation.rs:630]: fix #2079

**Review by Parth — APPROVED** [review]

**Pull request merged**